src/client/metric.rs

```diff
@@ -0,0 +1,528 @@
+use byteorder::WriteBytesExt;
+use memmap::{Mmap, MmapViewSync, Protection};
+use std::ffi::CString;
+use std::io;
+use std::io::{Cursor, Write};
+use std::mem;
+
+const ITEM_BIT_LEN: usize = 10;
+
+pub const I32_METRIC_TYPE_CODE: u32 = 0;
+pub const U32_METRIC_TYPE_CODE: u32 = 1;
+pub const I64_METRIC_TYPE_CODE: u32 = 2;
+pub const U64_METRIC_TYPE_CODE: u32 = 3;
+pub const F32_METRIC_TYPE_CODE: u32 = 4;
+pub const F64_METRIC_TYPE_CODE: u32 = 5;
+pub const STRING_METRIC_TYPE_CODE: u32 = 6;
+
+/// Generic type for any Metric's value
+pub trait MetricType {
+    /// Returns the MMV metric type code
+    fn type_code(&self) -> u32;
+    /// Writes the byte representation of the value to a writer.
+    ///
+    /// For integer and float types, the byte sequence is little endian.
+    ///
+    /// For the string type, the UTF-8 byte sequence is suffixed with a null byte.
+    fn write_to_writer<W: WriteBytesExt>(&self, writer: &mut W)
+        -> io::Result<()>;
+}
+
+macro_rules! impl_metric_type_for (
+    ($typ:tt, $base_typ:tt, $type_code:expr) => (
+        impl MetricType for $typ {
+            
+            fn type_code(&self) -> u32 {
+                $type_code
+            }
+
+            fn write_to_writer<W: WriteBytesExt>(&self, mut w: &mut W)
```

Comment:
It's already `&mut`, so you don't need `mut w` unless you were going to change what `w` is pointing to.
(Hmm, I'm surprised you're not getting an `unused_mut` warning...)